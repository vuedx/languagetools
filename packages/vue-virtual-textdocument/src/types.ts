export const TEMPLATE_BLOCK_SELECTOR = 'template'
export const SCRIPT_BLOCK_SELECTOR = 'script'
export const SCRIPT_SETUP_BLOCK_SELECTOR = 'scriptSetup'
export const RENDER_SELECTOR = '_render'
export const MODULE_SELECTOR = '_module'
export const INTERNAL_MODULE_SELECTOR = '_internal'
export const TEMPLATE_AST_SELECTOR = '_ast'

export type BlockSelector =
  | { type: typeof SCRIPT_BLOCK_SELECTOR }
  | { type: typeof SCRIPT_SETUP_BLOCK_SELECTOR }
  | { type: typeof TEMPLATE_BLOCK_SELECTOR }
  | { type: 'style'; index: number }
  | { type: 'customBlocks'; index: number }

export type Selector =
  | BlockSelector
  | { type: typeof RENDER_SELECTOR }
  | { type: typeof MODULE_SELECTOR }
  | { type: typeof INTERNAL_MODULE_SELECTOR }
  | { type: typeof TEMPLATE_AST_SELECTOR }

export type SelectorLike =
  | Selector
  | typeof TEMPLATE_BLOCK_SELECTOR
  | typeof SCRIPT_BLOCK_SELECTOR
  | typeof SCRIPT_SETUP_BLOCK_SELECTOR
  | typeof RENDER_SELECTOR
  | typeof MODULE_SELECTOR
  | typeof TEMPLATE_AST_SELECTOR
  | typeof INTERNAL_MODULE_SELECTOR
