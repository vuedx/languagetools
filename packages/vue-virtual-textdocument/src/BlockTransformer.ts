import type {
  CompilerError,
  RawSourceMap,
  SFCBlock,
  SFCDescriptor,
  SFCScriptBlock,
  SFCStyleBlock,
  SFCTemplateBlock,
} from '@vuedx/compiler-sfc'
import { annotations, compile } from '@vuedx/compiler-tsx'
import { pascalCase } from '@vuedx/shared'
import { isRootNode } from '@vuedx/template-ast-types'
import {
  createExportDeclarationForComponents,
  createExportDeclarationForDirectives,
  createExportDeclarationForExpose,
  createExportDeclarationForComponent,
  findScopeBindings,
  toAST,
  toCode,
} from '@vuedx/transforms'
import * as Path from 'path'
import type { VueSFCDocument } from './VueSFCDocument'

export { annotations }
export type Annotations = typeof annotations

export interface BlockTransformerOptions {
  block: SFCBlock | SFCScriptBlock | SFCStyleBlock | SFCTemplateBlock
  document: VueSFCDocument
  descriptor: SFCDescriptor
  annotations: Annotations
}

export interface TransformerError {
  message: string
  code: number | string | undefined
  codeDescription?: {
    href: string
  }
  start: number | undefined
  length: number | undefined
  source?: string
  severity: 'error' | 'warning' | 'info' | 'hint' | 'unused' | 'deprecated'
}

export type BlockTransformerFn = (
  source: string,
  id: string,
  options: BlockTransformerOptions,
) => {
  code: string
  map?: RawSourceMap
  errors?: TransformerError[]
  ast?: any
}

export interface BlockTransformer {
  transform: BlockTransformerFn
  output(block: SFCBlock): 'js' | 'jsx' | 'ts' | 'tsx'
}

function isSupportedLang(lang?: string): lang is 'js' | 'jsx' | 'ts' | 'tsx' {
  return ['js', 'jsx', 'ts', 'tsx'].includes(lang ?? '')
}

const caches = {
  script: new WeakMap<VueSFCDocument, string>(),
}

function toText(...lines: string[]): string {
  return lines.join('\n').replace(/\n$/, '') + '\n'
}

export const builtins: Record<'script' | 'template', BlockTransformer> = {
  script: {
    output: (block) => (isSupportedLang(block.lang) ? block.lang : 'js'),
    transform: (source, _id, { block, document }) => {
      try {
        const isScriptSetup =
          'setup' in block && block.setup != null && block.setup !== false
        // TODO: Should it be typescript instead of babel?
        const ast = toAST(source, {
          sourceFilename: document.fileName,
          isScriptSetup: isScriptSetup,
          lang: block.lang,
        })

        const usedIdentifiers = new Set<string>()
        const usedDirectives = new Set<string>()
        const usedComponents = new Set<string>()

        if (document.descriptor.template != null && isScriptSetup) {
          const ast = document.getDoc(document.descriptor.template)?.ast

          if (isRootNode(ast)) {
            ast.scope.globals.forEach((id) => {
              usedIdentifiers.add(id)
            })

            ast.components.forEach((name) => {
              const id = pascalCase(name.split('.').shift() ?? name)
              usedComponents.add(id)
            })

            ast.directives.forEach((name) => {
              const id = 'v' + pascalCase(name)
              usedDirectives.add(id)
            })
          }
        }

        const config = {
          isScriptSetup,
          leadingCommentForCopiedSource: annotations.copiedSource.start,
          trailingCommentForCopiedSource: annotations.copiedSource.end,
          leadingCommentForIdentifiers: annotations.templateGlobals.start,
          trailingCommentForIdentifiers: annotations.templateGlobals.end,
        }

        const nodes: any[] = [
          createExportDeclarationForComponents(ast, {
            ...config,
            shouldIncludeScriptSetup: (id) => usedComponents.has(id),
          }),
          createExportDeclarationForDirectives(ast, {
            ...config,
            shouldIncludeScriptSetup: (id) => usedDirectives.has(id),
          }),
          createExportDeclarationForExpose(ast, config),
        ]

        if (isScriptSetup) {
          document.declarations.identifiers = new Set(findScopeBindings(ast))

          nodes.push(
            createExportDeclarationForComponent(ast, {
              ...config,
              shouldIncludeScriptSetup: (id) => usedIdentifiers.has(id),
            }),
          )
        } else {
          nodes.push(createExportDeclarationForComponent(ast, config))
        }

        const code = toText(
          ';',
          annotations.diagnosticsIgnore.start,
          ';' + annotations.tsCompletions,
          toCode(nodes, { sourceText: source }).code,
          annotations.diagnosticsIgnore.end,
        )

        caches.script.set(document, code)

        return {
          ast,
          code: source + code,
        }
      } catch (error) {
        console.error('[VueDX] BlockTransformer - Error parsing script:', error)

        return {
          code:
            source +
            (caches.script.get(document) ??
              toText(
                ';',
                annotations.diagnosticsIgnore.start,
                ';' + annotations.tsCompletions,
                'export const __VueDX_components = {};',
                'export const __VueDX_directives = {};',
                'export const __VueDX_expose = {};',
                'export const __VueDX_DefineComponent = VueDX.internal.defineComponent({});',
                annotations.diagnosticsIgnore.end,
              )),
          errors: [],
        }
      }
    },
  },
  template: {
    output: () => 'tsx',
    transform: (source, _, { document, descriptor, block }) => {
      try {
        const selfSrcFileName = document.getBlockId(
          descriptor.scriptSetup ??
            descriptor.script ??
            document.fallbackScript,
        )

        const errors: CompilerError[] = []
        const result = compile(source, {
          sourceMap: true,
          filename: document.fileName,

          selfSrc: `./${Path.posix.basename(
            selfSrcFileName.substring(
              0,
              selfSrcFileName.length -
                Path.posix.extname(selfSrcFileName).length,
            ),
          )}`,

          onError(_e) {
            // errors.push(error)
          },
        })

        errors.push(...result.errors)
        document.templateAST = result.ast

        return {
          code: result.code,
          map: result.map,
          ast: result.ast,
          errors: errors.map((error) => {
            const diag: TransformerError = {
              message: error.message,
              code: error.code,
              severity: 'error',
              source: 'TemplateCompiler',
              start: undefined,
              length: undefined,
            }

            if (error.loc != null) {
              diag.start = error.loc.start.offset
              diag.length = error.loc.end.offset - error.loc.start.offset
            }

            return diag
          }),
        }
      } catch (e) {
        const error = e as Error
        return {
          code: '',

          errors: [
            {
              message: `Error processing template: ${String(
                error.message,
              )}\n${String(error.stack)}`,
              code: -1,
              source: 'TemplateCompiler',
              severity: 'error',
              start: 0,
              length: block.loc.source.length,
            },
          ],
        }
      }
    },
  },
}
