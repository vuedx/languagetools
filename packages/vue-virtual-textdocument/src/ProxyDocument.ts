import type {
  Position,
  Range,
  TextDocument,
} from 'vscode-languageserver-textdocument'

export abstract class ProxyDocument implements TextDocument {
  constructor(protected source: TextDocument) {}

  public get uri(): string {
    return this.source.uri
  }

  public get languageId(): string {
    return this.source.languageId
  }

  public get version(): number {
    return this.source.version
  }

  public get lineCount(): number {
    return this.source.lineCount
  }

  public getText(range?: Range): string {
    return this.source.getText(range)
  }

  public positionAt(offset: number): Position {
    return this.source.positionAt(offset)
  }

  public offsetAt(position: Position): number {
    return this.source.offsetAt(position)
  }
}
