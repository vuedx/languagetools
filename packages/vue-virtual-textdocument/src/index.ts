export {
  Position,
  Range,
  TextDocument,
  TextDocumentContentChangeEvent,
} from 'vscode-languageserver-textdocument'
export * from './VueSFCDocument'
