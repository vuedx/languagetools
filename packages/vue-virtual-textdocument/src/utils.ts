import { SFCBlock } from '@vuedx/compiler-sfc'
import { URI } from 'vscode-uri'
import { Selector } from './types'
import Path from 'path'

export function isNotNull<T>(value: T | null | undefined): value is T {
  return value != null
}

export function isString(value: unknown): value is string {
  return typeof value === 'string'
}

export function isNumber(value: unknown): value is number {
  return typeof value === 'number'
}

export function getLanguageIdFromExtension(ext: string) {
  switch (ext) {
    case 'js':
      return 'javascript'
    case 'ts':
      return 'typescript'
  }

  return ext
}

// Vue File: foo.vue
export const VIRTUAL_FILENAME_SEPARATOR = '________'

export function basename(fileName: string) {
  return Path.basename(fileName)
}

export function relativeVirtualImportPath(fileName: string) {
  return `./${basename(fileName).replace(/\.[\w]+$/, '')}`
}

export function isVueFile(fileName: string) {
  return fileName.endsWith('.vue')
}

export function isVirtualFile(fileName: string) {
  return fileName.indexOf('.vue' + VIRTUAL_FILENAME_SEPARATOR) > 0
}

export function getContainingFile(fileName: string) {
  return fileName.split(VIRTUAL_FILENAME_SEPARATOR).shift()!
}

export function asUri(fileName: string) {
  if (/^[a-z-]+:\/\//i.test(fileName)) return fileName

  const uri = URI.file(fileName).toString()

  if (isVirtualFile(fileName)) {
    return uri.replace(/^[^:]+/, 'vue')
  }

  return uri
}

export function asFilePath(path: string) {
  return URI.file(path).toString()
}

export function asFsPath(uri: string) {
  return URI.parse(uri).fsPath
}

export function parseVirtualFileName(fileName: string) {
  const uri = URI.parse(asUri(fileName))

  if (uri.scheme === 'vue') {
    const [container, selector] = uri.fsPath.split(VIRTUAL_FILENAME_SEPARATOR)
    if (!selector.includes('.'))
      throw new Error('Malformated virtual file uri.')
    const [block, index] = selector
      .substr(0, selector.lastIndexOf('.'))
      .split('__')

    return {
      uri: URI.file(container).toString(),
      selector: <Selector>(
        (index ? { type: block, index: parseInt(index, 10) } : { type: block })
      ),
    }
  }

  return null
}

const languages: Record<string, string> = {
  scriptSetup: 'javascript',
  script: 'javascript',
  template: 'vue-html',
  style: 'css',
  preview: 'vue-html',
  docs: 'markdown',
}

export function getBlockLanguage(block?: SFCBlock | null): string {
  if (!block) return ''

  if (block.lang) {
    if (block.lang === 'js') return 'javascript'
    if (block.lang === 'ts') return 'typescript'
    return block.lang
  }

  return languages[block.type] || block.type
}

export function isOffsetInBlock(
  offset: number,
  block?: SFCBlock | null,
): boolean {
  if (!block) return false

  return block.loc.start.offset <= offset && offset <= block.loc.end.offset
}

const extensions: Record<string, string> = {
  javascript: 'js',
  javascriptreact: 'jsx',
  typescript: 'ts',
  typescriptreact: 'tsx',
  markdown: 'md',
}
export function getLanguageExtension(lang: string): string {
  return extensions[lang] || lang
}

export function binarySearch<T>(
  array: T[],
  isMatch: (a: T) => number,
): T | undefined {
  let lo = 0
  let hi = array.length - 1

  while (lo <= hi) {
    const mid = ~~(lo + (hi - lo) / 2)
    const result = isMatch(array[mid])

    if (result === 0) {
      return array[mid]
    }
    if (result < 0) lo = mid + 1
    else hi = mid - 1
  }
}
