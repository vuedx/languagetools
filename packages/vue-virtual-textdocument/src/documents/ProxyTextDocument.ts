import {
  TextDocument,
  Position,
  Range,
} from 'vscode-languageserver-textdocument'
import { asFsPath, isNumber } from '../utils'

export abstract class ProxyTextDocument implements TextDocument {
  public readonly fsPath: string

  public constructor(protected doc: TextDocument) {
    this.fsPath = asFsPath(this.doc.uri)
  }

  public get version(): number {
    this.refresh()

    return this.doc.version
  }

  public get languageId(): string {
    return this.doc.languageId
  }

  public get lineCount(): number {
    this.refresh()

    return this.doc.lineCount
  }

  public get uri(): string {
    return this.doc.uri
  }

  public getText(): string
  public getText(range: Range): string
  public getText(offset: number, length: number): string
  public getText(range?: number | Range, length?: number): string {
    this.refresh()

    if (isNumber(range)) {
      return this.doc.getText().substr(range, length)
    } else {
      return this.doc.getText(range)
    }
  }

  public positionAt(offset: number): Position {
    this.refresh()

    return this.doc.positionAt(offset)
  }

  public offsetAt(position: Position): number {
    this.refresh()

    return this.doc.offsetAt(position)
  }

  protected refresh(): void {}
}
