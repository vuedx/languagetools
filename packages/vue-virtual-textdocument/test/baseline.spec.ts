import { sync as glob } from 'fast-glob'
import Path from 'path'
import FS from 'fs'
import { URI } from 'vscode-uri'
import { VueTextDocument } from '../src/documents/VueTextDocument'
import { asFilePath } from '../src/utils'

expect.addSnapshotSerializer({
  test: (value) => typeof value === 'string',
  print: (value) => `${value}`,
})

describe('VueVirtualDocument/baseline', () => {
  const dir = Path.resolve(__dirname, '../../../samples')
  const fileNames = glob('**/*.vue', {
    cwd: dir,
    absolute: false,
    ignore: ['node_modules'],
  })

  test.each(fileNames)('%s', async (name) => {
    const fileName = Path.resolve(dir, name)
    const content = await FS.promises.readFile(fileName, {
      encoding: 'utf-8',
    })
    const document = VueTextDocument.create(
      asFilePath(fileName),
      'vue',
      0,
      content,
    )

    expect(document.getDocument('_render')?.getText()).toMatchSnapshot()
  })
})
