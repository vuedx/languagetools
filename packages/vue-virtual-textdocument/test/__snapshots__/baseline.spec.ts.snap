// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VueVirtualDocument/baseline feature-diagnostics/src/FeaturePropType.vue 1`] = `
import _Ctx from './FeaturePropType.vue?internal'
import TypeScriptPropsType from './components/TypeScriptPropsType.vue'
import TypeScriptPropsOption from './components/TypeScriptPropsOption.vue'
import TypeScriptSetupPropsDeclare from './components/TypeScriptSetupPropsDeclare.vue'
import TypeScriptSetupPropsOption from './components/TypeScriptSetupPropsOption.vue'
import TypeScriptSetupPropsType from './components/TypeScriptSetupPropsType.vue'
import JavaScriptPropsOption from './components/JavaScriptPropsOption.vue'
import JavaScriptSetupPropsOption from './components/JavaScriptSetupPropsOption.vue'


export function render(_ctx: InstanceType<typeof _Ctx>) {
  return /*@@vue:start*/<><TypeScriptPropsType string="value" stringFull="value" number={1} numberFull={1} boolean booleanFull object={{ id: 1 }} objectFull={{ id: 1 }} array={[{ id: 1 }]} arrayFull={[{ id: 1 }]} union="string" unionFull={1}/><TypeScriptPropsOption string="value" stringFull="value" number={1} numberFull={1} boolean booleanFull object={{ id: 1 }} objectFull={{ id: 1 }} array={[{ id: 1 }]} arrayFull={[{ id: 1 }]} union="string" unionFull={1}/><TypeScriptSetupPropsDeclare string="value" stringFull="value" number={1} numberFull={1} boolean booleanFull object={{ id: 1 }} objectFull={{ id: 1 }} array={[{ id: 1 }]} arrayFull={[{ id: 1 }]} union="string" unionFull={1}/><TypeScriptSetupPropsOption string="value" stringFull="value" number={1} numberFull={1} boolean booleanFull object={{ id: 1 }} objectFull={{ id: 1 }} array={[{ id: 1 }]} arrayFull={[{ id: 1 }]} union="string" unionFull={1}/><TypeScriptSetupPropsType string="value" stringFull="value" number={1} numberFull={1} boolean booleanFull object={{ id: 1 }} objectFull={{ id: 1 }} array={[{ id: 1 }]} arrayFull={[{ id: 1 }]} union="string" unionFull={1}/><JavaScriptPropsOption string="value" stringFull="value" number={1} numberFull={1} boolean booleanFull object={{ id: 1 }} objectFull={{ id: 1 }} array={[{ id: 1 }]} arrayFull={[{ id: 1 }]} union="string" unionFull={1}/><JavaScriptSetupPropsOption string="value" stringFull="value" number={1} numberFull={1} boolean booleanFull object={{ id: 1 }} objectFull={{ id: 1 }} array={[{ id: 1 }]} arrayFull={[{ id: 1 }]} union="string" unionFull={1}/></>/*@@vue:end*/
}
`;

exports[`VueVirtualDocument/baseline feature-diagnostics/src/NoErrors.vue 1`] = `
import _Ctx from './NoErrors.vue?internal'
import TypeScriptPropsOption from './components/TypeScriptPropsOption.vue'


export function render(_ctx: InstanceType<typeof _Ctx>) {
  return /*@@vue:start*/<TypeScriptPropsOption string="value" stringFull="value" number={1} numberFull={1} boolean booleanFull object={{ id: 1 }} objectFull={{ id: 1 }} array={[{ id: 1 }]} arrayFull={[{ id: 1 }]} union="string" unionFull={1}/>/*@@vue:end*/
}
`;

exports[`VueVirtualDocument/baseline feature-diagnostics/src/components/JavaScriptPropsOption.vue 1`] = ``;

exports[`VueVirtualDocument/baseline feature-diagnostics/src/components/JavaScriptSetupPropsOption.vue 1`] = ``;

exports[`VueVirtualDocument/baseline feature-diagnostics/src/components/TypeScriptPropsOption.vue 1`] = ``;

exports[`VueVirtualDocument/baseline feature-diagnostics/src/components/TypeScriptPropsType.vue 1`] = ``;

exports[`VueVirtualDocument/baseline feature-diagnostics/src/components/TypeScriptSetupPropsDeclare.vue 1`] = ``;

exports[`VueVirtualDocument/baseline feature-diagnostics/src/components/TypeScriptSetupPropsOption.vue 1`] = ``;

exports[`VueVirtualDocument/baseline feature-diagnostics/src/components/TypeScriptSetupPropsType.vue 1`] = ``;

exports[`VueVirtualDocument/baseline feature-goto-definition/src/Javascript.vue 1`] = `
import _Ctx from './Javascript.vue?internal'
import HelloWorld from './components/HelloWorld.vue'


export function render({decrease, one, increase, two, name, email, code}: InstanceType<typeof _Ctx>) {
  return /*@@vue:start*/<><HelloWorld name="Jane"/><p><button onClick={decrease}>-</button> {one} <button onClick={increase}>+</button> and double is {two}</p><p>Name: {name} Email: {email} Code: {code}</p></>/*@@vue:end*/
}
`;

exports[`VueVirtualDocument/baseline feature-goto-definition/src/JavascriptSetup.vue 1`] = `
import _Ctx from './JavascriptSetup.vue?internal'
import HelloWorld from './components/HelloWorld.vue'


export function render({decrease, one, increase, two, name, email, code}: InstanceType<typeof _Ctx>) {
  return /*@@vue:start*/<><HelloWorld name="Jane"/><p><button onClick={decrease}>-</button> {one} <button onClick={increase}>+</button> and double is {two}</p><p>Name: {name} Email: {email} Code: {code}</p></>/*@@vue:end*/
}
`;

exports[`VueVirtualDocument/baseline feature-goto-definition/src/Typescript.vue 1`] = `
import _Ctx from './Typescript.vue?internal'
import HelloWorld from './components/HelloWorld.vue'


export function render({decrease, one, increase, two, name, email, code}: InstanceType<typeof _Ctx>) {
  return /*@@vue:start*/<><HelloWorld name="Jane"/><p><button onClick={decrease}>-</button> {one} <button onClick={increase}>+</button> and double is {two}</p><p>Name: {name} Email: {email} Code: {code}</p></>/*@@vue:end*/
}
`;

exports[`VueVirtualDocument/baseline feature-goto-definition/src/TypescriptSetup.vue 1`] = `
import _Ctx from './TypescriptSetup.vue?internal'
import HelloWorld from './components/HelloWorld.vue'


export function render({decrease, one, increase, two, name, email, code}: InstanceType<typeof _Ctx>) {
  return /*@@vue:start*/<><HelloWorld name="Jane"/><p><button onClick={decrease}>-</button> {one} <button onClick={increase}>+</button> and double is {two}</p><p>Name: {name} Email: {email} Code: {code}</p></>/*@@vue:end*/
}
`;

exports[`VueVirtualDocument/baseline feature-goto-definition/src/components/HelloWorld.vue 1`] = `
import _Ctx from './HelloWorld.vue?internal'


export function render({name}: InstanceType<typeof _Ctx>) {
  return /*@@vue:start*/<p>Hello {name}!</p>/*@@vue:end*/
}
`;

exports[`VueVirtualDocument/baseline feature-quickinfo/src/Javascript.vue 1`] = `
import _Ctx from './Javascript.vue?internal'
import HelloWorld from './components/HelloWorld.vue'


export function render({name, email, code, fullname, altCode}: InstanceType<typeof _Ctx>) {
  return /*@@vue:start*/<><HelloWorld name="Jane"/><p><span>Name: {name}</span><span>Email: {email}</span><span>Code: {code}</span></p><p><span>FullName: {fullname}</span><span>AltCode: {altCode}</span></p></>/*@@vue:end*/
}
`;

exports[`VueVirtualDocument/baseline feature-quickinfo/src/JavascriptSetup.vue 1`] = `
import _Ctx from './JavascriptSetup.vue?internal'
import HelloWorld from './components/HelloWorld.vue'


export function render({name, email, code, fullname, altCode}: InstanceType<typeof _Ctx>) {
  return /*@@vue:start*/<><HelloWorld name="Jane"/><p><span>Name: {name}</span><span>Email: {email}</span><span>Code: {code}</span></p><p><span>FullName: {fullname}</span><span>AltCode: {altCode}</span></p></>/*@@vue:end*/
}
`;

exports[`VueVirtualDocument/baseline feature-quickinfo/src/Typescript.vue 1`] = `
import _Ctx from './Typescript.vue?internal'
import HelloWorld from './components/HelloWorld.vue'


export function render({name, email, code, fullname, altCode}: InstanceType<typeof _Ctx>) {
  return /*@@vue:start*/<><HelloWorld name="Jane"/><p><span>Name: {name}</span><span>Email: {email}</span><span>Code: {code}</span></p><p><span>FullName: {fullname}</span><span>AltCode: {altCode}</span></p></>/*@@vue:end*/
}
`;

exports[`VueVirtualDocument/baseline feature-quickinfo/src/TypescriptSetup.vue 1`] = `
import _Ctx from './TypescriptSetup.vue?internal'
import HelloWorld from './components/HelloWorld.vue'


export function render({name, email, code, fullname, altCode}: InstanceType<typeof _Ctx>) {
  return /*@@vue:start*/<><HelloWorld name="Jane"/><p><span>Name: {name}</span><span>Email: {email}</span><span>Code: {code}</span></p><p><span>FullName: {fullname}</span><span>AltCode: {altCode}</span></p></>/*@@vue:end*/
}
`;

exports[`VueVirtualDocument/baseline feature-quickinfo/src/components/HelloWorld.vue 1`] = `
import _Ctx from './HelloWorld.vue?internal'


export function render({name}: InstanceType<typeof _Ctx>) {
  return /*@@vue:start*/<p>Hello {name}!</p>/*@@vue:end*/
}
`;

exports[`VueVirtualDocument/baseline feature-rename/src/Javascript.vue 1`] = `
import _Ctx from './Javascript.vue?internal'
import HelloWorld from './components/HelloWorld.vue'


export function render({decrease, one, increase, two, name, email, code}: InstanceType<typeof _Ctx>) {
  return /*@@vue:start*/<><HelloWorld name="Jane"/><p><button onClick={decrease}>-</button> {one} <button onClick={increase}>+</button> and double is {two}</p><p>Name: {name} Email: {email} Code: {code}</p></>/*@@vue:end*/
}
`;

exports[`VueVirtualDocument/baseline feature-rename/src/JavascriptSetup.vue 1`] = `
import _Ctx from './JavascriptSetup.vue?internal'
import HelloWorld from './components/HelloWorld.vue'


export function render({decrease, one, increase, two, name, email, code}: InstanceType<typeof _Ctx>) {
  return /*@@vue:start*/<><HelloWorld name="Jane"/><p><button onClick={decrease}>-</button> {one} <button onClick={increase}>+</button> and double is {two}</p><p>Name: {name} Email: {email} Code: {code}</p></>/*@@vue:end*/
}
`;

exports[`VueVirtualDocument/baseline feature-rename/src/Typescript.vue 1`] = `
import _Ctx from './Typescript.vue?internal'
import HelloWorld from './components/HelloWorld.vue'


export function render({decrease, one, increase, two, name, email, code}: InstanceType<typeof _Ctx>) {
  return /*@@vue:start*/<><HelloWorld name="Jane"/><p><button onClick={decrease}>-</button> {one} <button onClick={increase}>+</button> and double is {two}</p><p>Name: {name} Email: {email} Code: {code}</p></>/*@@vue:end*/
}
`;

exports[`VueVirtualDocument/baseline feature-rename/src/TypescriptSetup.vue 1`] = `
import _Ctx from './TypescriptSetup.vue?internal'
import HelloWorld from './components/HelloWorld.vue'


export function render({decrease, one, increase, two, name, email, code}: InstanceType<typeof _Ctx>) {
  return /*@@vue:start*/<><HelloWorld name="Jane"/><p><button onClick={decrease}>-</button> {one} <button onClick={increase}>+</button> and double is {two}</p><p>Name: {name} Email: {email} Code: {code}</p></>/*@@vue:end*/
}
`;

exports[`VueVirtualDocument/baseline feature-rename/src/components/HelloWorld.vue 1`] = `
import _Ctx from './HelloWorld.vue?internal'


export function render({name}: InstanceType<typeof _Ctx>) {
  return /*@@vue:start*/<p>Hello {name}!</p>/*@@vue:end*/
}
`;

exports[`VueVirtualDocument/baseline javascript/src/App.vue 1`] = `
import _Ctx from './App.vue?internal'
import HelloWorld from './components/HelloWorld.vue'


export function render(_ctx: InstanceType<typeof _Ctx>) {
  return /*@@vue:start*/<main><h1>A sample vue application</h1><HelloWorld name="Rahul"/></main>/*@@vue:end*/
}
`;

exports[`VueVirtualDocument/baseline javascript/src/components/HelloWorld.vue 1`] = `
import _Ctx from './HelloWorld.vue?internal'


export function render({name}: InstanceType<typeof _Ctx>) {
  return /*@@vue:start*/<p>Hello {name}!</p>/*@@vue:end*/
}
`;

exports[`VueVirtualDocument/baseline javascript-with-config/src/App.vue 1`] = `
import _Ctx from './App.vue?internal'
import HelloWorld from './components/HelloWorld.vue'


export function render(_ctx: InstanceType<typeof _Ctx>) {
  return /*@@vue:start*/<main><h1>A sample vue application</h1><HelloWorld name="Rahul"/></main>/*@@vue:end*/
}
`;

exports[`VueVirtualDocument/baseline javascript-with-config/src/components/HelloWorld.vue 1`] = `
import _Ctx from './HelloWorld.vue?internal'


export function render({name}: InstanceType<typeof _Ctx>) {
  return /*@@vue:start*/<p>Hello {name}!</p>/*@@vue:end*/
}
`;

exports[`VueVirtualDocument/baseline typescript/src/App.vue 1`] = `
import _Ctx from './App.vue?internal'
import HelloWorld from './components/HelloWorld.vue'


export function render(_ctx: InstanceType<typeof _Ctx>) {
  return /*@@vue:start*/<main><h1>A sample vue application</h1><HelloWorld name="Rahul"/></main>/*@@vue:end*/
}
`;

exports[`VueVirtualDocument/baseline typescript/src/components/HelloWorld.vue 1`] = `
import _Ctx from './HelloWorld.vue?internal'


export function render({name}: InstanceType<typeof _Ctx>) {
  return /*@@vue:start*/<p>Hello {name}!</p>/*@@vue:end*/
}
`;

exports[`VueVirtualDocument/baseline typescript-with-config/src/App.vue 1`] = `
import _Ctx from './App.vue?internal'
import HelloWorld from './components/HelloWorld.vue'


export function render(_ctx: InstanceType<typeof _Ctx>) {
  return /*@@vue:start*/<main><h1>A sample vue application</h1><HelloWorld name="Rahul"/></main>/*@@vue:end*/
}
`;

exports[`VueVirtualDocument/baseline typescript-with-config/src/components/HelloWorld.vue 1`] = `
import _Ctx from './HelloWorld.vue?internal'


export function render({name}: InstanceType<typeof _Ctx>) {
  return /*@@vue:start*/<p>Hello {name}!</p>/*@@vue:end*/
}
`;
